### PR TITLE
Update secp256k1 subtree to libsecp256k1 version 0.2.0

### PR DESCRIPTION
**Pull request by sipa**
Now that libsecp256k1 has a release (https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-December/021271.html), update the subtree to match it.

The changes themselves are not very impactful for Bitcoin Core, but include:
* It's no longer needed to specify whether contexts are for signing or verification or both (all contexts support everything), so make use of that in this PR.
* Verification operations can use the static context now, removing the need for some infrastructure in pubkey.cpp to make sure a context exists.
* Most modules are now enabled by default, so we can drop explicit enabling for them.
* CI improvements (in particular, MSVC and more recent MacOS)
* Introduction of an internal int128 type, which has no effect for GCC/Clang builds, but enables 128-bit multiplication in MSVC, giving a ~20% speedup there (but still slower than GCC/Clang).
* Release process changes (process documentation, changelog, ...).

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -2012,7 +2012,7 @@ LIBS_TEMP="$LIBS"
 unset LIBS
 LIBS="$LIBS_TEMP"
 
-ac_configure_args="${ac_configure_args} --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig"
```
Comment (Sjors) [inline configure.ac:2015]: Should be then explicitly disable some others?
https://github.com/bitcoin-core/secp256k1/commit/41e8704b484652cf5bbb2b7ecc27feedc3cf0ae1

`--disable-module-ecdh` and `--disable-module-extrakeys`? Or were these already enabled indirectly?

Comment (real-or-random) [reply configure.ac:2015]: `schnorrsig` depends on `extrakeys`, so `extrakeys` is automatically enabled already now.

`ecdh` is a prerequisite for https://github.com/bitcoin/bitcoin/pull/23561/ (see https://github.com/bitcoin/bitcoin/pull/23561/commits/240fd59a4817df1b3375c03d03dc0cbc9550530d) but I'm not able to tell if this is a reason to enable it right now.

File: src/pubkey.cpp [hunk 2]
```diff
@@ -388,7 +388,7 @@ ECCVerifyHandle::ECCVerifyHandle()
 {
     if (refcount == 0) {
         assert(secp256k1_context_verify == nullptr);
-        secp256k1_context_verify = secp256k1_context_create(SECP256K1_CONTEXT_VERIFY);
+        secp256k1_context_verify = secp256k1_context_create(SECP256K1_CONTEXT_NONE);
```
Comment (real-or-random) [inline src/pubkey.cpp:391]: We could equally use the static context here if we want to get rid of all the refcounting (but that's a somewhat larger change).

**Review by real-or-random — COMMENTED** [review]
Concept ACK

File: build_msvc/libsecp256k1/libsecp256k1.vcxproj [hunk 3]
```diff
@@ -14,7 +14,7 @@
   </ItemGroup>
   <ItemDefinitionGroup>
     <ClCompile>
-      <PreprocessorDefinitions>ENABLE_MODULE_ECDH;ENABLE_MODULE_RECOVERY;ENABLE_MODULE_EXTRAKEYS;ENABLE_MODULE_SCHNORRSIG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
Comment (Sjors) [inline build_msvc/libsecp256k1/libsecp256k1.vcxproj:17]: Do you need to explicitly disable it here too?

**Comment by real-or-random** [comment]
The third commit says "Invoke secp256k1_selftest() during initialization." but that happens in a different commit.

**Comment by Sjors** [comment]
I compiled 202291722300b86f36e97de7960d40a32544c2d1 on a Linux and an Intel mac, compiled and successfully ran the tests.

**Comment by fanquake** [comment]
cc @apoelstra @jonasnick

**Comment by Sjors** [comment]
For other reviewers, including myself, this should be useful for testing a subtree update: https://github.com/bitcoin/bitcoin/blob/master/test/lint/README.md#git-subtree-checksh

**Comment by Sjors** [comment]
ACK 202291722300b86f36e97de7960d40a32544c2d1, but 4462cb04986d77eddcfc6e8f75e04dc278a8147a could use more eyes on it.

File: build_msvc/libsecp256k1/libsecp256k1.vcxproj (hunk 3 above)
Comment (instagibbs) [inline build_msvc/libsecp256k1/libsecp256k1.vcxproj:17]: was ecdh enabled for any reason?

**Review by instagibbs — COMMENTED** [review]
commit message for 4462cb04986d77eddcfc6e8f75e04dc278a8147a

misnames `secp256k1_context_static` as `secp256k1_static_context`

**Comment by achow101** [comment]
ACK 202291722300b86f36e97de7960d40a32544c2d1

**Review by jonasnick — COMMENTED** [review]
utACK 202291722300b86f36e97de7960d40a32544c2d1

**Pull request merged**